### ISSUE TITLE
Q: Handling old-style inline attachments

### ISSUE THREAD

**Issue by rickhall** [mscdex/node-imap#598]
I have an message that is not multi-part, but has an inlined attachment that just comes after the message text like this:

...message text...

begin 666 ICSC 2008 Logos.pdf
M)5!$1BTQ+C0-)>+CS],-"C$P(# @;V)J#3P\+TQI;F5A<FEZ960@,2],(#$U
M-C8Q-B]/(#$R+T4@-S4Y-C(O3B R+U0@,34V,S8Y+T@@6R Q,#,V(#(T-5T^
M/@UE;F1O8FH-(" @(" @(" @(" @(" @#0IX<F5F#0HQ," S-PT*,# P,# P
M,# Q-B P,# P,"!N#0HP,# P,# Q,C@Q(# P,# P(&X-"C P,# P,#$S-3D@
...etc...

When viewed in Thunderbird, it detects this message as having an attachment. When I fetch the message metadata, I get the following structure:

[{"partID":"1","type":"text","subtype":"plain","params":null,"id":null,"description":null,"encoding":"7BIT","size":0,"lines":0,"md5":null,"disposition":null,"language":null}]

Up until this message, I had been simply fetching my message bodies by doing something like this:

    imap.fetch(uid, { 'bodies': [ '1' ] });

And this has always worked, except for this message, which results in a fetch error ("Some messages could not be FETCHed"). If I change bodies to simply be [''], then I can actually fetch the message body.

Two questions then:

1) Why can't I retrieve this message using '1' since the structure lists this partID? Is there something in the structure above so I can detect this situation?

2) To handle this old-style inline attachment, is this purely a client-side thing? In other words, I have no choice but to download the attachment (in the message body) and detect its there in the client?

Thanks.

**Comment by mscdex** [comment]
That doesn't seem to be the right structure if there is text in that message part, since the structure says the part is empty (`"size": 0`). Are you sure the structure fetch matches up with the body fetch?

**Comment by rickhall** [comment]
I don't have the exact fetch at hand right now, but I'm doing a fetch for some headers and "struct: true" just to get the structure of the message so I can determine which parts I want to fetch (I'm fetching only the message body, not the attachments). Would that account for why the size is 0?

This approach has been working for me up until this message. I was testing on one email account with 14k messages without issue, but now I started testing on an account with 180k messages. Then I discovered this email message.

So, in short, I'm pretty sure it matches up since its the way that I have been doing it. As I mentioned, if I don't request an explicit part and just request [''] then it does fetch the message.

**Comment by mscdex** [comment]
I think this may be by design. In general (as noted in the readme) the algorithm should be: check if the structure array's length > 1. If it is, use the `partId` to fetch the part, otherwise use `''` to fetch the body.

**Comment by rickhall** [comment]
Actually, I was thinking that would be my work around, but I didn't realize that was the recommended approach. So, I'll try that. Thanks. 

Regarding the attachment, should I just be scanning the message content on every message I retrieve on the client for the "begin"? 

I guess I only need to do this for single part text messages, because anything more modern probably wouldn't be sending attachments this way.

**Comment by mscdex** [comment]
As far as the "attachment", it looks to be [uuencoding](https://en.wikipedia.org/wiki/Uuencoding). As far as detecting it, you *might* be able to just use a regexp like:

```js
/\r?\nbegin (\d{3}) ([^\r\n]+)\r?\n([\s\S]+?)\r?\nend(?:$|\r?\n)/i
```

to check for and parse the file mode, file name, and uuencoded body.

**Comment by rickhall** [comment]
Yeah, I remember sending files that way on usenet. :-)

Thanks for the input.

**Comment by rickhall** [comment]
Just a quick follow up on this topic in case anyone else ends up here...

The fact that there was an uuencoded inlined attachment was a red herring with respect to why the message body could not be fetched. Apparently, this is related to some sort of faulty metadata in the Gmail server, similar to what is described here:

    https://bugzilla.mozilla.org/show_bug.cgi?id=510318

If I copied the offending message to a different folder, the copy appears to get its metadata updated correctly. After that, I'm able to request the body correctly even if I fetch it like:

    imap.fetch(uid, { 'bodies': [ '1' ] });

So, there you go. :-)

**Comment by mscdex** [comment]
@rickhall Thanks for the follow-up information!
